docs/randomized_testing.md

```diff
@@ -168,6 +168,250 @@ roughly the same things from the user (which operations to run, how to generate
 arguments to these operations, which failures to inject, which properties to
 check).
 
+## API
+
+In this section, we propose three options for a deterministic simulation API.
+Consider an application with components `A`, and `B`:
+
+```go
+// Component A
+type A interface { A(context.Context, int) (int, error) }
+type a struct{ weaver.Implements[A] }
+func (*a) A(context.Context, int) (int, error) { ... }
+
+// Component B
+type B interface { B(context.Context) error }
+type b struct{ weaver.Implements[B] }
+func (*b) B(context.Context) error { ... }
+
+// Component B fake.
+type fakeb struct{}
+func (*fakeb) B(context.Context) error { ... }
+```
+
+We want to simulate this application against a workload with the following
+characteristics:
+
+- We have two operations, or ops, to run. One calls `A.A` on randomly generated
+  integers, and the other calls `B.B`.
+- We want to fake component `B`.
+- We have some state that we update as the test runs. Specifically, we
+  want to record the set of all successful calls to `A.A`.
+
+### Option 1
+
+Option 1 is shown below. We call `sim.RegisterState[T]` to register a function
+that returns a freshly initialized piece of state of type `T`. We call
+`sim.RegisterFake[T]` to register a fake for component `T`. We call
+`sim.RegisterOp[T]` to register an operation. Every operation has
+
+- a name,
+- a way to randomly generate a value of type `T`, and
+- the body of the operation, which receives the state, the randomly
+  generated value of type `T`, and any components it needs to execute.
+
+Finally, we run the simulator. It's important to note that `s.Run()` runs many
+simulations, potentially on many threads. This is why `RegisterState` and
+`RegisterFake` take functions as arguments. The simulator has to create a new
+state and new fake for every simulation.
+
+```go
+func TestWorkload(t *testing.T) {
+    // Create the simulator.
+    s, err := sim.New(sim.Options{...})
+    if err != nil {
+        t.Fatal(err)
+    }
+
+    // Register state of type map[int]int.
+    sim.RegisterState(s, func() map[int]int {
+        return map[int]int{}
+    })
+
+    // Register a fake for component B.
+    sim.RegisterFake[B](s, func() B {
+        return &fakeb{}
+    })
+
+    // Register an op to call A.A.
+    sim.RegisterOp(s, sim.Op[int]{
+        Name: "CallA",
+        Gen:  func(*rand.Rand) int { ... },
+        Func: func(ctx context.Context, replies map[int]int, x int, a A) error {
+            if y, err := a.A(ctx, x); err == nil {
+                replies[x] = y
```

Comment:
I don't like this option because it's difficult to explain what instance of `replies` will be passed to the function. In reality, for each "workload", you'll be creating a state, and running a series of ops on it. But this isn't as encapsulated as it is in `Option 3`.

Reply:
I agree.